python/cudf/cudf/core/multiindex.py

```diff
@@ -178,18 +179,12 @@ def __init__(
                 )
 
         source_data = {}
-        for i, (column_name, col) in enumerate(codes._data.items()):
-            if -1 in col:
-                level = cudf.DataFrame(
-                    {column_name: [None] + list(levels[i])},
-                    index=range(-1, len(levels[i])),
-                )
-            else:
-                level = cudf.DataFrame({column_name: levels[i]})
-
-            source_data[column_name] = libcudf.copying.gather(
-                [level._data[column_name]], col
-            )[0]
+        for (column_name, col), level in zip(codes._data.items(), levels):
+            result_col = level._column.take(col)
```

Comment:
On line 175 we already do partial bounds-checking. The treatment of `-1` is AFAICT undocumented pandas behaviour. That is `codes` do not do normal indexing but instead `-1` is for "missing value".

But let us move all of the validation into one place:
```
import cudf._lib as libcudf
from cudf._lib.types import size_type_dtype
import numpy as np

lo, hi = libcudf.reduce.minmax(code)
if lo.value < -1 or hi.value > len(level) - 1:
    raise ValueError(...)
if lo.value == -1:
    # Now we can gather and insert null automatically
    code[code == -1] = np.iinfo(size_type_dtype).min
result_col, = libcudf.copying.gather([level], code, nullify=True)
```

Reply:
Nice find. Was able to consolidate the codes validate in this loop